packages/web-sdk/src/instrumentations/performance/navigation.ts

```diff
@@ -0,0 +1,48 @@
+import { genShortID } from '@grafana/faro-core';
+import type { EventsAPI } from '@grafana/faro-core';
+
+import { getItem, setItem, webStorageType } from '../../utils';
+
+import { NAVIGATION_ENTRY, NAVIGATION_ID_STORAGE_KEY } from './performanceConstants';
+import { calculateFaroNavigationTiming, calculateFaroResourceTiming, entryUrlIsIgnored } from './performanceUtils';
+import type { FaroNavigationItem } from './types';
+
+export function observeAndGetNavigationTimings(
+  pushEvent: EventsAPI['pushEvent'],
+  ignoredUrls: Array<string | RegExp>
+): Promise<FaroNavigationItem> {
+  let faroNavigationEntryResolve: (value: FaroNavigationItem) => void;
+  const faroNavigationEntryPromise = new Promise<FaroNavigationItem>((resolve) => {
+    faroNavigationEntryResolve = resolve;
+  });
+
+  const observer = new PerformanceObserver((observedEntries) => {
+    const [navigationEntryRaw] = observedEntries.getEntries();
+
+    if (navigationEntryRaw == null || entryUrlIsIgnored(ignoredUrls, navigationEntryRaw.name)) {
+      return;
+    }
+
+    const faroPreviousNavigationId = getItem(NAVIGATION_ID_STORAGE_KEY, webStorageType.session) ?? 'unknown';
+
+    const faroNavigationEntry: FaroNavigationItem = {
+      ...calculateFaroResourceTiming(navigationEntryRaw.toJSON()),
+      ...calculateFaroNavigationTiming(navigationEntryRaw.toJSON()),
+      faroNavigationId: genShortID(),
+      faroPreviousNavigationId,
+    };
+
+    setItem(NAVIGATION_ID_STORAGE_KEY, faroNavigationEntry.faroNavigationId, webStorageType.session);
```

Comment:
Note: I just realized we are [silencing](https://github.com/grafana/faro-web-sdk/blob/cf0375799c19529319cdc22d4b28c7a31a8651bc/packages/web-sdk/src/utils/webStorage.ts#L57) the errors from the storage. Might be worth taking care of this in the future

Reply:
We check if web storage avialable and usable before we call the storage api directly `isWebStorageTypeAvailable(webStorageMechanism)`.
For performance this function uses static values which are calculated when Faro is initialized.

```
function isWebStorageTypeAvailable(webStorageMechanism: StorageMechanism) {
  if (webStorageMechanism === webStorageType.local) {
    return isLocalStorageAvailable;
  }

  if (webStorageMechanism === webStorageType.session) {
    return isSessionStorageAvailable;
  }

  return false;
}
```

Of course there is a very small chance that in the meantime something happens like local|session storage is full etc.

But yes let's have look.